src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/VersionConverter.cs

```diff
@@ -1,25 +1,111 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
+using System.Buffers;
+using System.Buffers.Text;
+using System.Diagnostics;
+
 namespace System.Text.Json.Serialization.Converters
 {
     internal sealed class VersionConverter : JsonConverter<Version>
     {
+#if BUILDING_INBOX_LIBRARY
+        private const int MinimumVersionLength = 3; // 0.0
+
+        private const int MaximumVersionLength = 43; // 2147483647.2147483647.2147483647.2147483647
+
+        private const int MaximumEscapedVersionLength = JsonConstants.MaxExpansionFactorWhileEscaping * MaximumVersionLength;
+#endif
+
         public override Version Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
         {
+            if (reader.TokenType != JsonTokenType.String)
+            {
+                throw ThrowHelper.GetInvalidOperationException_ExpectedString(reader.TokenType);
+            }
+
+#if BUILDING_INBOX_LIBRARY
+            bool isEscaped = reader._stringHasEscaping;
+
+            int maxLength = isEscaped ? MaximumEscapedVersionLength : MaximumVersionLength;
+            ReadOnlySpan<byte> source = stackalloc byte[0];
+            if (reader.HasValueSequence)
+            {
+                if (!JsonHelpers.IsInRangeInclusive(reader.ValueSequence.Length, MinimumVersionLength, maxLength))
+                {
+                    throw ThrowHelper.GetFormatException(DataType.Version);
+                }
+
+                Span<byte> stackSpan = stackalloc byte[isEscaped ? MaximumEscapedVersionLength : MaximumVersionLength];
+                reader.ValueSequence.CopyTo(stackSpan);
+                source = stackSpan.Slice(0, (int)reader.ValueSequence.Length);
+            }
+            else
+            {
+                source = reader.ValueSpan;
+
+                if (!JsonHelpers.IsInRangeInclusive(source.Length, MinimumVersionLength, maxLength))
+                {
+                    throw ThrowHelper.GetFormatException(DataType.Version);
+                }
+            }
+
+            if (isEscaped)
+            {
+                int backslash = source.IndexOf(JsonConstants.BackSlash);
+                Debug.Assert(backslash != -1);
+
+                Span<byte> sourceUnescaped = stackalloc byte[MaximumEscapedVersionLength];
+
+                JsonReaderHelper.Unescape(source, sourceUnescaped, backslash, out int written);
+                Debug.Assert(written > 0);
+
+                source = sourceUnescaped.Slice(0, written);
+                Debug.Assert(!source.IsEmpty);
+            }
+
+            byte firstChar = source[0];
+            byte lastChar = source[source.Length - 1];
+            if (!JsonHelpers.IsDigit(firstChar) || !JsonHelpers.IsDigit(lastChar))
+            {
+                // Disallow leading and trailing whitespaces
+                // since Version.Parse allows them
```

Comment:
I don't understand this comment. If Version.Parse allows them, why disallow them here?

Reply:
Leading and trailing whitespaces are forbidden in System.Text.Json converters. See https://github.com/dotnet/runtime/pull/55564#issuecomment-926843371 and https://github.com/dotnet/runtime/pull/55564#discussion_r672594340
So we need to make sure to check input before passing it to Version.TryParse which silently parses it.
Wrote more explicit comment and copied it to .netstandard2.0 targeting version of S.T.J in 75c53e4